File: bindings/android/src/commonMain/kotlin/ai/uldren/loom/facets/columnar.kt
package ai.uldren.loom

    /** Create columnar dataset [name] with [columns] (CBOR array of `[name, type_tag]`) and [targetSegmentRows] (0 = default). */
expect fun Loom.columnarCreate(
        path: String,
        workspace: String,
        name: String,
        columns: ByteArray,
        targetSegmentRows: Long,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** Append [row] (a CBOR cell array) to columnar dataset [name] of [workspace]. */
expect fun Loom.columnarAppend(
        path: String,
        workspace: String,
        name: String,
        row: ByteArray,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** All rows of columnar dataset [name] in append order as a CBOR array of cell arrays. */
expect fun Loom.columnarScanCbor(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** The columns of dataset [name] as a CBOR array of `[name, type_tag]`. */
expect fun Loom.columnarColumnsCbor(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** The total row count of columnar dataset [name] of [workspace]. */
expect fun Loom.columnarRows(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): Long

    /** Compact columnar dataset [name] at its target segment size. */
expect fun Loom.columnarCompact(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    )


    /** Inspect columnar dataset [name] as CBOR `[columns, rows, segment_count, target_segment_rows, source_digest]`. */
expect fun Loom.columnarInspectCbor(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** Source digest for derived columnar projections as CBOR text. */
expect fun Loom.columnarSourceDigestCbor(
        path: String,
        workspace: String,
        name: String,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** Project [columns] (CBOR array of text) from rows matching CBOR [filter] (empty = all) as a CBOR array of cell arrays. */
expect fun Loom.columnarSelectCbor(
        path: String,
        workspace: String,
        name: String,
        columns: ByteArray,
        filter: ByteArray,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray


    /** Evaluate aggregate expressions from CBOR `[[op, column?] ...]`, with optional select [filter]. */
expect fun Loom.columnarAggregateCbor(
        path: String,
        workspace: String,
        name: String,
        aggregates: ByteArray,
        filter: ByteArray,
        passphrase: String? = null,
        kek: ByteArray? = null,
        authPrincipal: String? = null,
        authPassphrase: String? = null,
    ): ByteArray

File: bindings/react-native/src/facets/columnar.ts
import UldrenLoom from '../NativeUldrenLoom';
import { authArgs, keyArgs } from '../internal';
import type { LoomAuth, LoomKey } from '../internal';

/** Create columnar dataset `name` in `workspace` (created with the `columnar` facet if absent).
 * `columns` is raw Loom Canonical CBOR (an array of `[name, type_tag]`); `targetSegmentRows` is the
 * segment size (0 for the default). */
export function columnarCreate(
  loomPath: string,
  workspace: string,
  name: string,
  columns: Uint8Array | number[],
  targetSegmentRows: number,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.columnarCreate(
    loomPath,
    workspace,
    name,
    Array.from(columns),
    targetSegmentRows,
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

/** Append `row` (raw Loom Canonical CBOR, a cell array) to dataset `name`, validating arity and column
 * types. */
export function columnarAppend(
  loomPath: string,
  workspace: string,
  name: string,
  row: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.columnarAppend(
    loomPath,
    workspace,
    name,
    Array.from(row),
    passphrase,
    kek,
    authPrincipal,
    authPassphrase
  );
}

/** All rows of dataset `name` in append order as raw Loom Canonical CBOR (an array of cell arrays). */
export async function columnarScanCbor(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.columnarScan(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase)
  );
}

/** The `(name, type_tag)` columns of dataset `name` as raw Loom Canonical CBOR (an array of
 * `[name, type_tag]`). */
export async function columnarColumnsCbor(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.columnarColumns(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase)
  );
}

/** The total row count of dataset `name`. */
export function columnarRows(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<number> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.columnarRows(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase);
}

/** Compact dataset `name` at its target segment size. */
export function columnarCompact(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<void> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return UldrenLoom.columnarCompact(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase);
}

/** Inspect dataset metadata as raw Loom Canonical CBOR. */
export async function columnarInspectCbor(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.columnarInspect(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase)
  );
}

/** Source digest used by derived columnar projections as raw Loom Canonical CBOR text. */
export async function columnarSourceDigestCbor(
  loomPath: string,
  workspace: string,
  name: string,
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.columnarSourceDigest(loomPath, workspace, name, passphrase, kek, authPrincipal, authPassphrase)
  );
}

/** Project `columns` (raw Loom Canonical CBOR, an array of text) from rows of dataset `name` matching
 * `filter` (raw Loom Canonical CBOR `[col, op, value_cell]`; empty for all), as raw Loom Canonical
 * CBOR (an array of cell arrays). */
export async function columnarSelectCbor(
  loomPath: string,
  workspace: string,
  name: string,
  columns: Uint8Array | number[],
  filter: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.columnarSelect(
      loomPath,
      workspace,
      name,
      Array.from(columns),
      Array.from(filter),
      passphrase,
      kek,
      authPrincipal,
      authPassphrase
    )
  );
}

/** Evaluate aggregate expressions from raw Loom Canonical CBOR `[[op, column?] ...]`, with optional filter. */
export async function columnarAggregateCbor(
  loomPath: string,
  workspace: string,
  name: string,
  aggregates: Uint8Array | number[],
  filter: Uint8Array | number[],
  loomKey?: LoomKey,
  auth?: LoomAuth
): Promise<Uint8Array> {
  const [passphrase, kek] = keyArgs(loomKey);
  const [authPrincipal, authPassphrase] = authArgs(auth);
  return Uint8Array.from(
    await UldrenLoom.columnarAggregate(
      loomPath,
      workspace,
      name,
      Array.from(aggregates),
      Array.from(filter),
      passphrase,
      kek,
      authPrincipal,
      authPassphrase
    )
  );
}
